### ISSUE TITLE
FTPClient.storeFile() deferreds interact badly (passive mode)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8058]
|<img alt="fdrake's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| fdrake reported|
|-|-|
|Trac ID|trac#433|
|Type|defect|
|Created|2003-12-13 07:47:42Z|

Attachments:

* [storefile.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/82a/82ad38f885211232bd89c439e0df9982d6ae74f7/ae9de65ff325c69b3d87bc96ac2adc061cba4bac.py) (2229 bytes) - added by fdrake on 2003-12-13 08:56:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__433 433
type__defect defect
reporter__fdrake fdrake
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1071301662000000 1071301662000000
changetime__1469663957305933 1469663957305933
version__ 
owner__ 
cc__spiv cc__fdrake
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="fdrake's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|fdrake commented|
|-|-|

```
#!html
<pre>
The deferred and deferred list returned by FTPClient.storeFile() don't interact
in the way I expect in the case when the control connection refuses to allow the
upload.  My expectation was that if the control connection responds with an
error (causing the deferred list's errback to be called), the callback for the
consumer deferred will not have been won't be call (not sure how I'd want it's
errback to be handled; I wanted to call it from the errback for the deferred list).

What I want to do is produce a single deferred that succeeds when the control
connection says everything is OK and the data connection is established.  If
either fails, the errback of my deferred should be called.

I think this warrants fixing, but I may have to live with Twisted 1.1.1 as
released at this point.  I'll follow up as I make progress.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="fdrake's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|fdrake commented|
|-|-|

```
#!html
<pre>
I've attached a file that contains the essence of my work-around.  This approach
can be used with Twisted 1.1.1 as released.  My application contains more
logging and such, but really nothing more than what's here.  This is not a
complete script, but I think it shows what I did effectively.

The approach is to wait until both the deferred and deferred list returned by
FTPClient.storeFile succeed, and then fire the callbacks on another deferred. 
In the case of an error, the "outer" deferred's errback() is called instead.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
(This is really just a general debugging tip, rather than specific to this
issue...)

On IRC, you mentioned that you were getting AlreadyCalledErrors.  If you set
Deferred.debug = True, Deferred's will have 'creator' and 'invoker'
attributes that record the tracebacks of when they were instantiated and
first called, which can be very handy to figure out exactly what's
happening.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I've finally gotten around to looking at this bug again...  sorry for taking so
long!

> What I want to do is produce a single deferred that succeeds when the control
> connection says everything is OK and the data connection is established.  If
> either fails, the errback of my deferred should be called.

Isn't that what DeferredList is for? 

e.g.:

    dConsumer, dResponses = ftpClient.storeFile(path)
    d = DeferredList([dConsumer, dResponses], fireOnOneErrback=True)
    
    def doWhatever((ok1, consumer), (ok2, responses)):
        do_something_with(consumer)

    d.addCallback(doWhatever)
    return d

Your workaround seems to be just a less general form of DeferredList.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
In the absence of an explanation of why DeferredList with fireOnOneErrback is
inadequate, I'm marking this as resolved.

Please feel free to reopen this if you disagree!
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|GitHub commented|
|-|-|

In changeset a7dcf6639157c2d68436c57bbadb3827280670be
```
#!CommitTicketReference repository="" revision="a7dcf6639157c2d68436c57bbadb3827280670be"
Merge pull request #433 from twisted/8697-einarfd-py3-conch-client-knownhosts

Author: einarfd
Reviewer: rodrigc
Fixes: #8697

Port twisted.conch.client.knownhosts to Python 3
```
